Modules/RosterManager/Roster.lua

```diff
@@ -83,6 +87,53 @@ function Roster:Standings(GUID)
     end
 end
 
+function Roster:WeeklyGains(GUID, week)
```

Comment:
Doing it your way would just limit the functionality of this function or require additional checking in the code.  Please propose structure for this function to show what you have in mind.

Reply:
```lua
function Roster:GetAllWeeklyGains()
    return self.weeklyGains or {}
end
function Roster:GetWeeklyGainsForPlayer(GUID)
    Utils.assertString(GUID, 'GUID')
    return self.weeklyGains[GUID] or {}
end
function Roster:GetWeeklyGainsForPlayerWeek(GUID, week)
    Utils.assertString(GUID, 'GUID')
    Utils.assertInteger(GUID, 'week')
    return self.weeklyGains[GUID][week] or 0
end
```

To me this code is way more readable. You go from 9 lines to 12 lines; but:
- Complexity is significantly reduced.
- The extra lines now do full parameter type verification with proper error handling
- Each function now has a single return type instead of one that depends on the number of arguments given.

Also, for the caller of the function this is way safer. If I intend to call the original code but my week number is `nil` for some reason I get a dataset with a different structure / type (table instead of number), this will probably make my code fail somewhere in the future.
In the new code I get an error immediately because we each function has 1 task and can verify that it has received the proper arguments for doing so.

FYI: The `assertX` functions are not yet in CLM Utils, but are simple to implement (and reusable). The reason for passing in the name of the argument is for better error messages, it could be omitted.

Reply:
Applied in newest commit